packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Spec](./README.md)
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text data-key="name"></nimble-table-column-text>
+    <nimble-table-column-icon data-key="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation. The ordering of the columns in the markup will determine the visual ordering of the columns (top to bottom equals left to right...unless in 'rtl'). Re-ordering of columns will be done, at least at first, through the re-ordering of the columns elements in the DOM.
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    /*
+     * @internal
+     */
+    @observable
+    public readonly slottedColumns: TableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where the `TCellData` type is a superset of the type represented by [`TableRecord`](https://github.com/ni/nimble/blob/3e4b8d3dd59431d1671e381aa66052db57bc475c/packages/nimble-components/src/table/types.ts#L24):
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
```

Comment:
At one point during tech stack research we had talked about whether cells might need to be able to serialize some state when they are about to be scrolled off-screen to be retreived and deserialized before they are scrolled back on-screen. (As a contrived example, this could be something like the scroll position of a text field in a cell.)

Would a client use TableCellState to store this type of information? If so, what's the lifecycle for when the table requests it?

Or is that out of scope for this proposal?

Reply:
So, I don't think a client even _could_ use the `TableCellState` to store that type of information, in that clients only have access to that type via a custom column implementation, and that object is only for static configuration. Dynamic state that needs to be maintained _could_ be done via the `TableCellState` interface, but it would be updated and accessed solely through the `Table` implementation (i.e. hidden from the client, and thus the state of interest to maintain would be decided by the `Table` itself).

For now, I would propose that the details for how we handle the situation @jattasNI raises are tabled, if we can agree that the current APIs are not prohibitive (which I believe they are not).

Reply:
Just to update this comment a thread on a new working decision, column element implementations will no longer generate a `TableCellState` instance, and instead only return column configuration state that is needed by the `cellTemplate` implementation. This moves the column element API even further from the concern @jattasNI raises.